### DISCUSSION TITLE
REST API

### DISCUSSION THREAD

**Discussion by BatD2** [PaloAltoNetworks/pan-os-python#438] [Ideas]
Is there any plans to integrate the Palo Alto REST API to the pan-os-python framwork? 

I also use the XML API in most of the cases, but the REST and XML  API seem to be behaving differently when working with Panorama objects. For example when showing address objects in a device group, XML API retuns only the device group specific address objects, while the REST returns all available to the group address objects, including obejcts inherited from parent groups. 
This can be very useful for specific usecases and it will be useful to have the option to choose between the xml and rest.

**Comment by shinmog** [comment]
There are no current undertakings to support for REST API currently.  It still has gaps that only the XML API can do, such as logging in, op commands, and commits, just to name a few.

The XML API can give you all that information, you just have to build your object hierarchy appropriately.  Discovering those hierarchies is as simple as using [hierarchy_info()](https://pan-os-python.readthedocs.io/en/latest/useful-methods.html#informational-methods).

**Comment by BatD2** [comment]
Ok, so certainly you can pull objects, find hierarchy, then presuming that there are 3 parent groups, build the rule order from firmst to last for per-rule and the oposite for post rules. Then repeat for all other rule types. 
On the other hand it is a single API call to get all this. I am not suggesting replacing XML with teh REST api , but adding the rest api fucitonality option.
